Add support for foreignObject and enhance label editing with multi-line support

planger opened the pull request:
https://github.com/eclipse/sprotty/issues/169

planger commented:
![Peek 2020-03-24 09-36](https://user-images.githubusercontent.com/588090/77405207-2248ee80-6db3-11ea-8011-4cb32f8886a4.gif)

tomvdbussche commented on src/lib/generic-views.tsx:
```diff
@@ -35,3 +38,26 @@ export class PreRenderedView implements IView {
+        </g>;
+        setAttr(node, 'class', model.type);
+        const subType = getSubType(model);
+        if (subType) setAttr(node, 'class', subType);
```
The subtype is added for every view in `CssClassPostprocessor` so it's not be needed here.

tomvdbussche commented on src/features/edit/edit-label-ui.ts:
```diff
@@ -193,24 +210,33 @@ export class EditLabelUI extends AbstractUIExtension {
+            x = bounds.x + (this.label.editControlBoundsCorrection ? this.label.editControlBoundsCorrection.x * zoom : 0);
+            y = bounds.y + (this.label.editControlBoundsCorrection ? this.label.editControlBoundsCorrection.y * zoom : 0);
+            height = bounds.height + (this.label.editControlBoundsCorrection ? this.label.editControlBoundsCorrection.height * zoom : 0);
+            width = bounds.width + (this.label.editControlBoundsCorrection ? this.label.editControlBoundsCorrection.width * zoom : 0);
```
These changes have some unintended side-effects for vertical editable labels (see edges in the class diagram example)

planger replied:
Thanks, good catch! This is actually a tough one...
Previously, this just didn't cause problems "by accident" because we added a quite generous additional width and didn't apply the height to the input but only to the container (change below `this.editControl.style.height ...`).
We don't really know whether the label is rotated here. I guess the most pragmatic approach is to set a fixed height and width (as before this change) and let the label implementation decide if it wants to set height and width in a specific way (e.g. to the bounds). I'll make a suggestion in a new commit very soon.

tomvdbussche replied:
That makes sense

planger commented:
Thanks a lot @tomvdbussche for your review! I hope https://github.com/eclipse/sprotty/pull/171/commits/70252d490de07800e4fb90e34d0c74b0c9451e80 addresses your comments.
